File: Infrastructure/CubicBezierEase.vb
Imports System
Imports System.Windows.Media.Animation

Namespace Infrastructure
    Public Class CubicBezierEase
        Implements IEasingFunction

        Private Const NewtonIterations As Integer = 8
        Private Const BinarySubdivisionIterations As Integer = 10
        Private Const Precision As Double = 0.000001

        Public Property X1 As Double
        Public Property Y1 As Double
        Public Property X2 As Double
        Public Property Y2 As Double

        Public Function Ease(normalizedTime As Double) As Double Implements IEasingFunction.Ease
            If normalizedTime <= 0.0 Then
                Return 0.0
            End If

            If normalizedTime >= 1.0 Then
                Return 1.0
            End If

            Dim parameter = SolveCurveX(normalizedTime)
            Return SampleCurveY(parameter)
        End Function

        Private Function SolveCurveX(targetX As Double) As Double
            Dim estimate = targetX

            For iteration = 0 To NewtonIterations - 1
                Dim currentX = SampleCurveX(estimate) - targetX
                If Math.Abs(currentX) < Precision Then
                    Return estimate
                End If

                Dim derivative = SampleCurveDerivativeX(estimate)
                If Math.Abs(derivative) < Precision Then
                    Exit For
                End If

                estimate -= currentX / derivative
            Next

            Dim lower = 0.0
            Dim upper = 1.0
            estimate = targetX

            For iteration = 0 To BinarySubdivisionIterations - 1
                Dim currentX = SampleCurveX(estimate)
                If Math.Abs(currentX - targetX) < Precision Then
                    Exit For
                End If

                If currentX < targetX Then
                    lower = estimate
                Else
                    upper = estimate
                End If

                estimate = (lower + upper) * 0.5
            Next

            Return estimate
        End Function

        Private Function SampleCurveX(parameter As Double) As Double
            Return ((A(X1, X2) * parameter + B(X1, X2)) * parameter + C(X1)) * parameter
        End Function

        Private Function SampleCurveY(parameter As Double) As Double
            Return ((A(Y1, Y2) * parameter + B(Y1, Y2)) * parameter + C(Y1)) * parameter
        End Function

        Private Function SampleCurveDerivativeX(parameter As Double) As Double
            Return (3.0 * A(X1, X2) * parameter + 2.0 * B(X1, X2)) * parameter + C(X1)
        End Function

        Private Shared Function A(first As Double, second As Double) As Double
            Return 1.0 - 3.0 * second + 3.0 * first
        End Function

        Private Shared Function B(first As Double, second As Double) As Double
            Return 3.0 * second - 6.0 * first
        End Function

        Private Shared Function C(first As Double) As Double
            Return 3.0 * first
        End Function
    End Class
End Namespace

File: src/ClickSyncMouseTester/Infrastructure/CubicBezierEase.cs
using System;
using System.Windows.Media.Animation;

namespace ClickSyncMouseTester.Infrastructure;

public class CubicBezierEase : IEasingFunction
{
    private const int NewtonIterations = 8;

    private const int BinarySubdivisionIterations = 10;

    private const double Precision = 1E-06;

    public double X1 { get; set; }

    public double Y1 { get; set; }

    public double X2 { get; set; }

    public double Y2 { get; set; }

    public double Ease(double normalizedTime)
    {
        if (normalizedTime <= 0.0)
        {
            return 0.0;
        }
        if (normalizedTime >= 1.0)
        {
            return 1.0;
        }

        double curveParameter = SolveCurveX(normalizedTime);
        return SampleCurveY(curveParameter);
    }

    double IEasingFunction.Ease(double normalizedTime)
    {
        return Ease(normalizedTime);
    }

    private double SolveCurveX(double targetX)
    {
        double parameter = targetX;
        for (int iteration = 0; iteration < NewtonIterations; iteration++)
        {
            double xError = SampleCurveX(parameter) - targetX;
            if (Math.Abs(xError) < Precision)
            {
                return parameter;
            }

            double derivative = SampleCurveDerivativeX(parameter);
            if (Math.Abs(derivative) < Precision)
            {
                break;
            }

            parameter -= xError / derivative;
        }

        double lowerBound = 0.0;
        double upperBound = 1.0;
        parameter = targetX;
        for (int iteration = 0; iteration < BinarySubdivisionIterations; iteration++)
        {
            double sampledX = SampleCurveX(parameter);
            if (Math.Abs(sampledX - targetX) < Precision)
            {
                break;
            }

            if (sampledX < targetX)
            {
                lowerBound = parameter;
            }
            else
            {
                upperBound = parameter;
            }
            parameter = (lowerBound + upperBound) * 0.5;
        }
        return parameter;
    }

    private double SampleCurveX(double parameter)
    {
        return ((A(X1, X2) * parameter + B(X1, X2)) * parameter + C(X1)) * parameter;
    }

    private double SampleCurveY(double parameter)
    {
        return ((A(Y1, Y2) * parameter + B(Y1, Y2)) * parameter + C(Y1)) * parameter;
    }

    private double SampleCurveDerivativeX(double parameter)
    {
        return (3.0 * A(X1, X2) * parameter + 2.0 * B(X1, X2)) * parameter + C(X1);
    }

    private static double A(double first, double second)
    {
        return 1.0 - 3.0 * second + 3.0 * first;
    }

    private static double B(double first, double second)
    {
        return 3.0 * second - 6.0 * first;
    }

    private static double C(double first)
    {
        return 3.0 * first;
    }
}
